### DISCUSSION TITLE
Is there a way to check if `OutputPath` has been overwritten in a csproj?

### DISCUSSION THREAD

**Discussion by thecristidima** [dotnet/msbuild#8113] [Q&A]
I want to group most of my projects in a few output folders and I'm setting `OutputPath` in `Directory.Build.targets` using some simple conditions (e.g. test projects go to `TestOutput`). I also have a few projects that I would like to exclude from these rules as they have a custom `OutputPath` already set in their `.csproj`.

I was wondering if there is a recommended way of checking if `OutputPath` was overwritten in a `.csproj`. The only solution I found at the moment is to set a custom property instead of `OutputPath` in the `.csproj` (e.g. `CustomOutputPath`) and when that property is set, I use it instead of the generic rules I defined. Another variation of this would be just using that second property as a flag and keeping `OutputPath` as it is.

Is there a simpler/cleaner way of doing this? Ideally, I would like to be able to just add some kind of condition like `'($OutputPath)' != '($DefaultOutputPath)', but I don't think this exists (or at least I haven't found it yet).

**Comment by rainersigwald** [comment]
I can't think of a really good way; I like your `$(CustomOutputPath)` solution.

You could maybe try to compute if `$(OutputPath)` matches our default computations:

https://github.com/dotnet/msbuild/blob/3777dcaf7edb3e86a070037ba53e742dd1872873/src/Tasks/Microsoft.Common.CurrentVersion.targets#L154-L156

But that sounds pretty fragile.

**Comment by KalleOlaviNiemitalo** [comment]
I hoped .NET SDK might provide a way to do this:

- make Directory.Build.props set a property CustomBeforeMicrosoftNETDefaultOutputPathsTargets = MyDefaultPaths.targets
- make .NET SDK import $(CustomBeforeMicrosoftNETDefaultOutputPathsTargets) at the start of Microsoft.NET.DefaultOutputPaths.targets, or in a similar place
- in MyDefaultPaths.targets, set OutputPath if empty

However it looks like .NET SDK does not recognise a property like that. There are other similar properties but SDK imports the referenced files too late.

**Comment by JanKrivanek** [comment]
Closing as currently there is no better solution.
Feel free to open feature request if you feel this is important and needs better handling.
